### PR TITLE
Fix and improve RMNodeUpdater

### PR DESCRIPTION
**Pull request by fviale**
- fork a new JVM to execute the up-to-date node.jar (otherwise, there are classpath conflicts)
 - forward properties from the current JVM to the new one (filtering system properties)
 - clean the One-Jar expand dir (to prevent jar duplicates coming from different versions).
 - add binlib to the OneJar forked JVM classpath (only way to allow powershell script engine execution)

### CODE REVIEW

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 1]
```diff
@@ -123,36 +145,185 @@ private static boolean makeNodeUpToDate() {
...
         }
+    }
 
-        return false;
+    /**
+     * Clean the directory used by One-Jar to expand libraries. After a node.jar update, we clean this directory to prevent jar conflicts
+     * @param jarFile name of the node jar file
+     * @throws IOException
+     */
+    private static void cleanExpandDirectory(String jarFile) throws IOException {
```
Comment (sonarqube-activeeon) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: MINOR Remove the declaration of thrown exception 'java.io.IOException', as it cannot be thrown from method's body. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 2]
```diff
@@ -67,8 +89,8 @@ private static boolean isLocalJarUpToDate(String url, String filePath) {
             System.out.println("Url date=" + new Date(urlConnection.getLastModified()));
             System.out.println("File date=" + new Date(file.lastModified()));
 
-            if (file.lastModified() < urlConnection.getLastModified()) {
-                System.out.println("Local jar " + file + " is obsolete");
+            if (!file.exists() || file.lastModified() < urlConnection.getLastModified()) {
+                System.out.println("Local jar " + file + " is obsolete or not present");
```
Comment (sonarqube-activeeon) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: MAJOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java (hunk 2 above)
Comment (marcocast) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: System.out?

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 3]
```diff
@@ -101,7 +123,7 @@ private static boolean makeNodeUpToDate() {
 
                     if (destination.exists()) {
 
-                        lockFile = new File(System.getProperty(TMP_DIR_PROPERTY) + "/lock");
+                        lockFile = new File(StandardSystemProperty.JAVA_IO_TMPDIR.value(), "lock");
```
Comment (marcocast) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: you don t need the  / before the lock ?

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java (hunk 1 above)
Comment (marcocast) [reply rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: as mr sonar said....

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 4]
```diff
@@ -123,36 +145,185 @@ private static boolean makeNodeUpToDate() {
...
+     * Clean the directory used by One-Jar to expand libraries. After a node.jar update, we clean this directory to prevent jar conflicts
+     * @param jarFile name of the node jar file
+     * @throws IOException
+     */
+    private static void cleanExpandDirectory(String jarFile) throws IOException {
+        File directoryToClean;
+        String oneJarExpandDir = System.getProperty(ONEJAR_EXPAND_DIR_PROPERTY);
+        if (oneJarExpandDir == null) {
+            // Default scheme used by one-jar
+            String jar = new File(jarFile).getName().replaceFirst("\\.[^\\.]*$", "");
```
Comment (marcocast) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: could you put this regex in a variable with a humanly readable name? like FIRST_SLASH = "\\.[^\\.]*$"

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 5]
```diff
@@ -123,36 +145,185 @@ private static boolean makeNodeUpToDate() {
...
+        }
+
+        FileUtils.deleteQuietly(directoryToClean);
     }
 
-    public static void main(String[] args) throws IOException, InterruptedException, ClassNotFoundException,
-            NoSuchMethodException, InvocationTargetException, IllegalAccessException {
-        makeNodeUpToDate();
+    public static void main(String[] args) throws Exception {
+        while (!makeNodeUpToDate()) {
```
Comment (marcocast) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: no timeout?

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java (hunk 3 above)
Comment (fviale) [reply rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: not when I use new File(parent, child)

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java (hunk 4 above)
Comment (fviale) [reply rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: ok

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java (hunk 5 above)
Comment (fviale) [reply rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: you are right, I will add a timeout

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java (hunk 2 above)
Comment (fviale) [reply rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: Ok I will add loggers

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 6]
```diff
@@ -46,113 +74,496 @@
...
+                        logger.info("Releasing the lock on " + lockFile.getAbsoluteFile());
+                        lock.release();
+                        channel.close();
+
+                        return false;
+                    }
+                }
+                fetchUrl(nodeJarUrl, destination);
+                // Align the local file modification time with the remote url.
+                destination.setLastModified(getRemoteLastModifiedMillis(nodeJarUrl));
```
Comment (sonarqube-activeeon) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: MINOR Do something with the "boolean" value returned by "setLastModified". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS899)

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 7]
```diff
@@ -46,113 +74,496 @@
...
+    private static void trustEveryone() {
+        try {
+            HttpsURLConnection.setDefaultHostnameVerifier(new HostnameVerifier() {
+                public boolean verify(String hostname, SSLSession session) {
+                    return true;
+                }
+            });
+            SSLContext context = SSLContext.getInstance("TLS");
+            context.init(null, new X509TrustManager[] { new X509TrustManager() {
+                public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
```
Comment (sonarqube-activeeon) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 8]
```diff
@@ -46,113 +74,496 @@
...
+                }
+
+                public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
```
Comment (sonarqube-activeeon) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 9]
```diff
@@ -46,113 +74,496 @@
...
+        try {
+            cl = parser.parse(options, args);
+            //now we update this object's fields given the options.
+            String nodeName = fillParameters(cl, options);
+            //check the user supplied values
+            //performed after fillParameters to be able to override fillParameters in subclasses
+            checkUserSuppliedParameters();
+            return nodeName;
+        } catch (ParseException pe) {
+            pe.printStackTrace();
```
Comment (sonarqube-activeeon) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: MINOR Use a logger to log this exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1148)

File: rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java [hunk 10]
```diff
@@ -46,113 +74,496 @@
...
+
+    public static final String LAST_DOT_AND_AFTER = "\\.[^\\.]*$";
+
+    private static final Logger logger = initLogger();
 
-    private static final String TMP_DIR_PROPERTY = "java.io.tmpdir";
+    public static final int NUMBER_OF_ATTEMPTS = 10;
 
-    public static final String ONE_JAR_JAR_PATH = "one-jar.jar.path";
+    public RMNodeUpdater() {
```
Comment (sonarqube-activeeon) [inline rm/rm-node-updater/src/main/java/org/ow2/proactive/resourcemanager/updater/RMNodeUpdater.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by fviale** [comment]
Retest this please

**Pull request merged**